### DISCUSSION TITLE
Automatic formula calculation

### DISCUSSION THREAD

**Discussion by chirag-compro** [exceljs/exceljs#2507] [Ideas & Proposals]
#41 
Can we have a utility which automatically calculates the formula and updates the **result** key.
Requirement for browser version.

**Comment by mfahmirukman** [comment]
can you show the code?

**Comment by chirag-compro** [comment]
Usecase for above issue.
Value in A1 =5;
Value in C1 =6;
If I update the value of cell **B1** .
  worksheet.getCell('B1').value = { formula 'A1+C1'}; 
Now If I retrieve the value of cell B1.
 worksheet.getCell('B1').value
Actual Result: { formula 'A1+C1', result:undefined}
Expected Result: { formula 'A1+C1', result:11}

**Comment by Rycochet** [comment]
That would require implementing the entire Excel parser and interpreter to support...

**Comment by mfahmirukman** [comment]
Hmm, at this point even after reading the documentation I actually get confused as to why I need to use the formula at all. The formula expects me to know what the result is per documentation.

> Note that ExcelJS cannot process the formula to generate a result, it must be supplied.

I'm not sure whether this is usable during the processing of file writing/stream writing it. After I finished writing the file, I opened it in Microsoft Excel and cell B1 showed the result of the formula

_**I've only done summing numbers, I haven't tried multiplying or dividing. But I'll do it once I'm back at my workplace. I'm only speculating that this might also work with other forms of calculation**_

 The way I do it for now is just how I'd do it in Microsoft Excel
say I want to sum 2 cells, that being cell A1 and A2 and put the result in B1
value in A1 = 5
value in A2 = 7


I'd fill B1 with a string like this:
`worksheet.getCell('B1').value = "=SUM(A1,A2)"`
Edit: My bad, I actually forgot whether it was this line above or this one below
`worksheet.getCell('B1').value = { formula: "SUM(A1,A2)"}`

instead of doing this `worksheet.getCell('B1').value = { formula 'A1+C1'};`

**Comment by mfahmirukman** [comment]
Tried what you meant by _undefined_

`sheet.getCell('B1').value = {formula : 'SUM(A1,A2)'}`
`{ formula: 'SUM(A1,A2)' }`
`sheet.getCell('B1').value`
`{ formula: 'SUM(A1,A2)', result: undefined }`

I tried another way by doing this 
`sheet.getCell('B1').value = {formula : 'SUM(A1,A2)', result: (sheet.getCell('A1').value + sheet.getCell('A2').value) }`

and got this

`{ formula: 'SUM(A1,A2)', result: 10 }`

**Comment by chirag-compro** [comment]
@Rycochet  Yes, I believe it includes a complete implementation of the formula parser for Excel. But will help in retaining the JSON structure of workbook and further use of the workbook.
@mfahmirukman The last implemetation you have provided will require to write the parser for calculating the result key.

**Comment by mfahmirukman** [comment]
@chirag-compro  yes you're correct. If you're talking about doing the calculation to get the result during the process of the file writing then I'd have to write the parser first.

But, in most cases, I've only used this instead. It's true that I can't get the result key with this during the filewriting as seen above but if I at least know the Excel formula and what I want to calculate, I can let Microsoft Excel parse the result.
`worksheet.getCell('B1').value = { formula: "SUM(A1,A2)"} `

**Comment by chirag-compro** [comment]
@mfahmirukman  Yes, it is just an extra feature ExcelJS can provide in near future. It would increase the capability of ExcelJS , as the user can now retain the JSON for future use instead of relying on the Microsoft Excel Parser for the formulas.

**Comment by mah110020** [comment]
While it is true that you can specify a formula without a result, and that Excel will parse the result,
`worksheet.getCell("A2").value = { formula: "A1+B1" }`
I wanted to add that this is not the case when the document is opened in protected view, such as when it is downloaded from the internet. In this case, the cell will appear empty.

**Comment by giordifungula** [comment]
has this issue been resovled ? it still looks like you need to know the value or sum of the forumula still. Anyone find any aid @mah110020  @chirag-compro  @mfahmirukman

**Comment by mah110020** [comment]
There isn't any built in parser. You might be able to pull one in from
other libraries, but it's often not necessary. Because the workbook can be
configured to have excel recalculate all the results when initially loaded.

`workbook.calcProperties.fullCalcOnLoad = true;`

This obviously doesn't work if opened by an excel view-only app, or if
opened in read-only/protected mode.

**Comment by giordifungula** [comment]
@mah110020 Okay, do you know the best tool to assist with parsing this and if you don't mind why is parsing needed for the excel sheet ?  will the file be corrupted or something ?

**Comment by Siemienik** [comment]
ExcelJs doesn't support counting formulas, 
It is able to set `value: undefined`, which force recalculating in a MS Excel when enable file to edit (turnoff readonly mode).

Recalculating in JavaScript needs to implement OpenFormula standard. The best tool which do this and I know it is HyperFormula: https://github.com/handsontable/hyperformula (obviously read the license)

**Comment by krzcho** [comment]
What if I have a file which stubbornly only gives formulas without results. I am not able to force Excel to calculate them :(
